### PR TITLE
Feature 3 : Edit / Update ToDo functionality implemented.

### PR DESCRIPTION
**Pull request by saa-wtag**
[Task-3: Edit/Update Todo](https://trello.com/c/qMw8fYHA)
This PR introduces the functionality of Edit/Update a todo by clicking the edit button beside input box. If any user doesn't want to change the todo-title by clicking the cancel button, the previous state will be visible.

**Functionalities:**

-  Beside each todo, there is another new button named Edit.
-  Clicking the Edit todo button will the turn the current Todo text content into an input box.
-  When in Edit Mode, The text of Add button is changed from Edit to Update.
-  If a user clicked the edit button, but don’t want any changes, then there should be a button named cancel to go back to the previous state.
- If input is unchanged or empty then update button will remain disable.
- While editing one task if we click another task's edit button then previous task's text box will collapse and current task's text box will get open.

Please have a close look on this PR and kindly put your valuable comments.

https://github.com/sad-wtag/ToDoApp_VanillaJS/assets/164306406/35364430-77bf-4530-9c30-3ecce8ced7b2

### CODE REVIEW

File: js/script.js [hunk 1]
```diff
@@ -26,6 +27,63 @@ const createTask = (taskTitle)=>{
...
+    if (currentTaskId !== null && currentTaskId !== taskId) {
+        cancelEdit();
+    }
+
+    currentTaskId = taskId;
+
+    const taskIndex = tasks.findIndex((task) => task.id === taskId);
+    const currentTask = tasks[taskIndex];
+    const newId = new Date().getTime();
+    console.log(taskId);
```
Comment (aka-wtag) [inline js/script.js]: remove this console.log

Comment (saa-wtag) [reply js/script.js]: @aka-wtag bhai I have removed unwanted lines and executed prettier to format the codes. Please can you review once again.
Thank you

File: js/script.js [hunk 2]
```diff
@@ -17,33 +21,71 @@ const deleteHandler = (taskId) => {
...
+const editHandler = (task) => {
+  cancelEdit();
+  task.editMode = true;
+  renderTasks();
+};
+
+const updateHandler = (task, newTitle) => {
+  if (newTitle.trim().length > 0) {
+    task.title = newTitle.trim();
+  }
```
Comment (aka-wtag) [inline js/script.js]:
```suggestion
  const title = newTitle.trim();
  if (title) {
    task.title = title;
  }
```

**Review by aka-wtag — APPROVED** [review]

File: js/script.js [hunk 3]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+    showToastMessage("Please provide a valid title!");
+  }
 };
 
 const deleteTask = (taskId) => {
   tasks = tasks.filter((task) => task.id !== taskId);
   renderTasks();
 };
 
+const editHandler = (task) => {
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
const editTask = (task) => {
```

File: js/script.js [hunk 4]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+  cancelEdit();
+  task.editMode = true;
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
  task.isEditing = true;
```

File: js/script.js [hunk 5]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+  renderTasks();
+};
+
+const updateHandler = (task, newTitle) => {
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
const updateTasks = (task, newTitle) => {
```

File: js/script.js (hunk 2 above)
Comment (maa-wtag) [reply js/script.js]:
```suggestion
  // Check if the new title is not empty after trimming whitespace
if (newTitle.trim()) {
    // Update the task title with the trimmed version of the new title
    task.title = newTitle.trim();
}
```

File: js/script.js [hunk 6]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+  renderTasks();
+};
+
 const createTask = (taskTitle) => {
   const task = {
     id: new Date().getTime(),
     title: taskTitle,
     createdAt: formatDate(new Date()),
     deleteButton: createElement("Delete", "button", () => deleteTask(task.id)),
+    editMode: false,
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
    isEditing: false,
```

File: js/script.js [hunk 7]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
   $taskList.innerHTML = "";
-
   tasks.forEach((task) => {
-    const texts = [task.title, " Created at: " + task.createdAt];
-    const $tasksList = createElement(texts, "li");
+    const $tasksList = document.createElement("li");
+    const $titleElement = document.createElement("span");
+    $titleElement.textContent = task.title;
+
+    if (task.editMode) {
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
    if (task.isEditing) {
```

File: js/script.js [hunk 8]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+      const $inputField = document.createElement("input");
+      $inputField.type = "text";
+      $inputField.value = task.title;
+
+      const $updateButton = createElement("Update", "button", () =>
+        updateHandler(task, $inputField.value)
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
        updateTask(task, $inputField.value)
```

File: js/script.js [hunk 9]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+      );
+      const $cancelButton = createElement("Cancel", "button", cancelEdit);
+
+      $tasksList.append($inputField, $updateButton, $cancelButton);
+    } else {
+      const $deleteButton = createElement("Delete", "button", () =>
+        deleteTask(task.id)
+      );
+      const $editButton = createElement("Edit", "button", () =>
+        editHandler(task)
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
        isEditing(task)
```

File: js/script.js [hunk 10]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
 };
 
 const renderTasks = () => {
   $taskList.innerHTML = "";
-
   tasks.forEach((task) => {
-    const texts = [task.title, " Created at: " + task.createdAt];
-    const $tasksList = createElement(texts, "li");
+    const $tasksList = document.createElement("li");
+    const $titleElement = document.createElement("span");
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
    const $taskTitle = document.createElement("span");
```

File: js/script.js [hunk 11]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+    $titleElement.textContent = task.title;
```
Comment (maa-wtag) [inline js/script.js]: ditto

File: js/script.js [hunk 12]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+      $tasksList.append($inputField, $updateButton, $cancelButton);
+    } else {
+      const $deleteButton = createElement("Delete", "button", () =>
+        deleteTask(task.id)
+      );
+      const $editButton = createElement("Edit", "button", () =>
+        editHandler(task)
+      );
+
+      $tasksList.append($titleElement, $deleteButton, $editButton);
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
      $tasksList.append($taskTitle, $deleteButton, $editButton);
```

File: js/script.js [hunk 13]
```diff
@@ -14,40 +14,81 @@ const addButtonHandler = () => {
...
+    }
 
-    $tasksList.appendChild(task.deleteButton);
     $taskList.appendChild($tasksList);
   });
 };
 
+const cancelEdit = () => {
+  tasks.forEach((task) => (task.editMode = false));
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
  tasks?.forEach((task) => (task.isEditing = false));
```

**Review by maa-wtag — APPROVED** [review]
@saa-wtag Please resolve the comments
also please resolve the conflicting files before proceeding to the next step

File: js/script.js (hunk 2 above)
Comment (saa-wtag) [reply js/script.js]: Now, I have sanitized the updatedTitle before passing it to the handler function. I hope now the issue is resolved.

File: js/script.js (hunk 13 above)
Comment (saa-wtag) [reply js/script.js]: Done

File: js/script.js (hunk 5 above)
Comment (saa-wtag) [reply js/script.js]: Done

File: js/script.js (hunk 4 above)
Comment (saa-wtag) [reply js/script.js]: Done

File: js/script.js (hunk 3 above)
Comment (saa-wtag) [reply js/script.js]: Done

File: .vscode/settings.json
Comment (hhu-wtag) [inline .vscode/settings.json:1]: This is not related to the project for now. Let's add this file to .gitignore

File: js/script.js [hunk 14]
```diff
@@ -14,36 +14,62 @@ const addButtonHandler = () => {
...
+      updateTask,
+      cancelEdit
+    );
     $taskList.appendChild($taskElement);
   });
 };
 
+const cancelEdit = (curTask) => {
+  tasks.forEach((task) => {
+    if (task.id === curTask.id) task.isEditing = false;
```
Comment (hhu-wtag) [inline js/script.js]:
```suggestion
    if (task.id === curTask.id){
        task.isEditing = false;
    }
```

**Review by hhu-wtag — APPROVED** [review]
